src/gc.c

```diff
@@ -344,6 +345,16 @@ static void run_finalizers(void)
     JL_GC_POP();
 }
 
+void jl_gc_inhibit_finalizers(int state)
+{
+    if (jl_gc_finalizers_inhibited && !state && !jl_in_gc) {
+        jl_in_gc = 1;
```

Comment:
the problem with running finalizers during codegen is that we may be currently trying to generate code for that function or some other required function when the finalizer tries to run. this patch re-introduces that possibility, but just makes it slightly rarer.

for inference, we can check the `in-inference` flag and if set, we instead run a non-inferred version. the equivalent that would be required here is to the `in-compile` flag and run a non-compiled version of the function.

i'm not sure of the issue with recursively running finalizers, since that has been allowed in the past.

the `jl_in_gc` flag was introduced to inhibit calls to yield inside finalizers, which would allow fairly unpredictable task switching to occur